README.md
# moxa-irigb-driver

The IRIG-B driver is for moxa DA-820C embedded computer.
`/dev/moxa_irigb` is the entry point controlling the IRIG-B device.

1. Setting the path to your kernel source tree in Makefile.
To compile this driver, you should modify the KDIR to the path of your kernel source.
```
KDIR:=/lib/modules/`uname -r`/build
KERNEL_VERSION=`uname -r`
```

2. Compile the driver
```bash
root@Moxa:# make
```


3. Install and load the driver
You can install the driver
```bash
root@Moxa:# make install
```

Or install the driver manually
```bash
root@Moxa:# mkdir /lib/modules/`uname -r`/kernel/drivers/misc
root@Moxa:# cp -a moxa_irigb.ko /lib/modules/`uname -r`/kernel/drivers/misc/
```

and Load the driver manually
```bash
root@Moxa: insmod /lib/modules/`uname -r`/kernel/drivers/misc/moxa_irigb.ko
```

4. Load the driver in the initial script
```bash
load_irigb_driver() {
    if [ "`lsmod|grep -c moxa_irigb`" = "0" ]; then
        # If the driver has not been loaded, load it
        insmod /lib/modules/`uname -r`/kernel/drivers/misc/moxa_irigb.ko
    fi
}
```

6. IRIG-B tools
[moxa-irigb-tools](http://gitlab.syssw.moxa.com/MXcore-Package/moxa-irigb-tools)


debian/changelog
moxa-irigb-driver (1.1.0) stretch; urgency=low

  * update dependence to Debian 9.12
  * add postinst for depmod

 -- Elvis Yao (姚家瑋) <ElvisCW.Yao@moxa.com>  Fri, 17 Apr 2020 19:01:26 +0800

moxa-irigb-driver (1.0.0) stretch; urgency=low

  * Initial release.

 -- Elvis Yao (姚家瑋) <ElvisCW.Yao@moxa.com>  Thu, 30 Jan 2020 17:21:07 +0800


moxa_irigb.c
/*  Copyright (C) MOXA Inc. All rights reserved.
 *
 *      This software is distributed under the terms of the
 *          MOXA License.  See the file COPYING-MOXA for details.
 *
 *     File name: moxa_irigb.c
 *
 *	  Device file:
 *	  	mknod /dev/IRIG-B c 10 $minor
 *
 *	v1.0	08/26/2014 Jared Wu	New release for DA-820-LX IRIG-B
 *	v1.1	11/04/2014 Jared Wu	Suport Redhat Enterprise 5.6, linux-2.6.18
 *	v1.2	05/05/2015 Jared Wu	Fix the got too many interrupts while IRIG-B transist to off-line status.
 *	v1.3	18/10/2019 Elvis Yao	Fix coding style
 *
 */
#include <linux/version.h>
#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/init.h>
#include <linux/fs.h>
#include <linux/mm.h>
#include <linux/pci.h>
#include <linux/ioport.h>
#include <linux/wait.h>
#include <linux/poll.h>
#include <linux/cdev.h>
#include <linux/delay.h>
#include <linux/device.h>
#include <linux/miscdevice.h>
#include <linux/spinlock.h>
#include <linux/interrupt.h>
#include "moxa_irigb.h"

#define DRIVER_VERSION		"1.3"

struct irigb_struct {
	unsigned long phyaddr;
	unsigned long phylen;
	void *vaddr;
	unsigned int irq;
	unsigned char di_state_keep;		/* current di state */
	unsigned char do_state_keep;		/* current do state */
	struct pci_dev *pdev;			/* pci device */
	struct miscdevice irigb_misc;		/* struct miscdevice */
	u32 TimeStatus;
	spinlock_t counter_lock;		/* lock for counter state */
} ;

/* Define the function prototype for IRIG-B */
int irigb_probe(struct pci_dev *pdev, const struct pci_device_id *ent);
void irigb_remove(struct pci_dev *pdev);

int irigb_open (struct inode *inode, struct file *fp);
long irigb_ioctl (struct file *fp, unsigned int cmd, unsigned long arg);
ssize_t irigb_read(struct file *fp, char *buffer, size_t length, loff_t *offset);
int irigb_release (struct inode *inode, struct file *fp);

/* Defined the golbal variables */
/* This variable is for IRIG_B */
struct irigb_struct IRIG_B;

static struct file_operations irigb_fops = {
	.owner = THIS_MODULE,
	.open = irigb_open,
	.unlocked_ioctl = irigb_ioctl,
	.read = irigb_read,
	.release = irigb_release,
};

static struct pci_device_id irigb_pcibrds[] = {
	{PCI_VENDOR_ID_MOXA, MX_IRIGB_DEVICE_ID, PCI_ANY_ID, PCI_ANY_ID, 0, 0, (kernel_ulong_t *)&IRIG_B},
	{0}
};

static struct pci_driver irigb_driver = {
	.name = MX_IRIGB_NAME,
	.id_table = irigb_pcibrds,
	.probe = irigb_probe,
	.remove = irigb_remove
};

int irigb_open(struct inode *inode, struct file *fp)
{
	return 0;
}

ssize_t irigb_read(struct file *fp, char *buffer, size_t length, loff_t *offset)
{
	return 0;
}

long irigb_ioctl (struct file *fp, unsigned int cmd, unsigned long arg) {
	long retval = 0;
	struct reg_val_pair_struct reg_val_pair;
	struct reg_bit_pair_struct reg_bit_pair;
	u32 *reg = IRIG_B.vaddr;
	int i;
	u32 val;

	switch (cmd) {
	case IOCTL_GET_REGISTER:
		retval = copy_from_user(&reg_val_pair, (struct reg_val_pair_struct *)arg, sizeof(struct reg_val_pair_struct));
		if (retval < 0) {
			printk(KERN_INFO"irigb_ioctl(): IOCTL_GET_REGISTER copy_from_user() fail\n");
			return retval;
		}
		if (reg_val_pair.count == 0) {
			printk(KERN_INFO"irigb_ioctl(): IOCTL_GET_REGISTER reg_val_pair.count is 0\n");
			retval = -EINVAL;
			return retval;
		}

		for (i = 0; i < reg_val_pair.count; i++) {
			reg_val_pair.val[i] = (u32) readl(reg + reg_val_pair.addr[i]);
		}

		retval = copy_to_user((struct reg_val_pair_struct *)arg, &reg_val_pair, sizeof(struct reg_val_pair_struct));
		if (retval < 0) {
			printk(KERN_INFO"irigb_ioctl(): IOCTL_GET_REGISTER copy_to_user() fail\n");
			return retval;
		}
		break;
	case IOCTL_SET_REGISTER:
		retval = copy_from_user(&reg_val_pair, (struct reg_val_pair_struct *)arg, sizeof(struct reg_val_pair_struct));
		if (retval < 0) {
			printk(KERN_INFO"irigb_ioctl(): IOCTL_SET_REGISTER copy_from_user() fail\n");
			return retval;
		}

		if (reg_val_pair.count == 0) {
			printk(KERN_INFO"irigb_ioctl(): IOCTL_SET_REGISTER reg_val_pair.count is 0\n");
			retval = -EINVAL;
			return retval;
		}
		for (i = 0; i < reg_val_pair.count; i++) {
			writel(reg_val_pair.val[i], reg + reg_val_pair.addr[i]);
		}
		break;
	case IOCTL_SETCLR_REGISTER_BIT:
		retval = copy_from_user(&reg_bit_pair, (struct reg_bit_pair_struct *)arg, sizeof(struct reg_bit_pair_struct));
		if (retval < 0) {
			printk(KERN_INFO"irigb_ioctl(): IOCTL_SETCLR_REGISTER_BIT copy_from_user() fail\n");
			return retval;
		}
		val = (u32) readl(reg + reg_bit_pair.addr);
		val &= (~reg_bit_pair.clear_bit);
		val |= (reg_bit_pair.set_bit);
		writel(val, reg + reg_bit_pair.addr);
		break;
	case IOCTL_GET_TIMESRC_STATUS:
		retval = copy_to_user((u32 *)arg, &IRIG_B.TimeStatus, sizeof(u32));
		if (retval < 0) {
			printk(KERN_INFO"irigb_ioctl(): IOCTL_GET_TIMESRC_STATUS copy_to_user() fail\n");
			return retval;
		}
		break;
	default:
		printk("ioctl:error\n");
		retval = -EINVAL;
		return retval ;
	}

	return retval;
}

int irigb_release(struct inode *inode, struct file *fp)
{
	return 0;
}

static irqreturn_t irigb_interrupt(int irq, void *dev_id)
{
	u32 intMsk;
	u32 intSts;
	u32 maskReg;
	struct irigb_struct *pIRIG_B = (struct irigb_struct *) dev_id;
	u32 *reg = pIRIG_B->vaddr;

	intMsk = ~(readl(reg+INTMSK));

	if (intMsk == 0) {
		return IRQ_NONE;
	}

	intSts = readl(reg+INTSTS);
	if ((intMsk & intSts) == 0) {
		return IRQ_NONE;
	}

	maskReg = INTSTS_BIT_IRIG0DE_OFF | INTSTS_BIT_IRIG0DE_FRMERR | INTSTS_BIT_IRIG0DE_PARERR | INTSTS_BIT_IRIG0DE_DONE;
	if (0 != (intSts & maskReg)) {
		pIRIG_B->TimeStatus &= (~maskReg);
		pIRIG_B->TimeStatus |= (intSts & maskReg);

		/* mask INTSTS_BIT_IRIG0DE_OFF to avoid interrupt every 20ms */
		if (intSts & maskReg & INTSTS_BIT_IRIG0DE_OFF)
			intMsk &= ~INTSTS_BIT_IRIG0DE_OFF;
		else
			intMsk |= INTSTS_BIT_IRIG0DE_OFF;
	}

	maskReg = INTSTS_BIT_IRIG1DE_OFF|INTSTS_BIT_IRIG1DE_FRMERR|INTSTS_BIT_IRIG1DE_PARERR|INTSTS_BIT_IRIG1DE_DONE;

	if (0 != (intSts & maskReg)) {
		pIRIG_B->TimeStatus &= (~maskReg);
		pIRIG_B->TimeStatus |= (intSts & maskReg);

		/* mask INTSTS_BIT_IRIG1DE_OFF to avoid interrupt every 20ms */
		if (intSts & maskReg & INTSTS_BIT_IRIG1DE_OFF)
			intMsk &= (~INTSTS_BIT_IRIG1DE_OFF);
		else
			intMsk |= INTSTS_BIT_IRIG1DE_OFF;
	}

	maskReg = INTSTS_BIT_PPSDE_TIMEOUT | INTSTS_BIT_PPSDE_DONE;

	if (0 != (intSts & maskReg)) {
		pIRIG_B->TimeStatus &= ~maskReg;
		pIRIG_B->TimeStatus |= (intSts&maskReg);

		/* mask INTSTS_BIT_PPSDE_TIMEOUT to avoid interrupt every 20ms */
		if (intSts & maskReg & INTSTS_BIT_PPSDE_TIMEOUT)
			intMsk &= (~INTSTS_BIT_PPSDE_TIMEOUT);
		else
			intMsk |= INTSTS_BIT_PPSDE_TIMEOUT;
	}
#if 0	/* Jared: remove to prevent got too many interrupts while IRIG-B got off-line */
	intMsk = readl(reg+INTMSK);
#endif
	writel(~intMsk, reg+INTMSK);

	return IRQ_HANDLED;
}

int irigb_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
{
	int retval = 0;
	u32 intMsk = 0xFFFFFFFF;
	struct resource *r;
	struct irigb_struct *pIRIG_B = &IRIG_B;
	u32 *reg;

	printk(KERN_INFO"Found device %x:%x\n", pdev->vendor, pdev->device);

	pci_set_drvdata(pdev, &IRIG_B);

	pIRIG_B->pdev = pdev;

	pIRIG_B->phyaddr = pci_resource_start(pdev, 0);
	pIRIG_B->phylen = pci_resource_len(pdev, 0);

	r = request_mem_region(pIRIG_B->phyaddr, pIRIG_B->phylen, MX_IRIGB_NAME);
	if (r == NULL) {
		printk("Moxa %s request_mem_region() fail !\n", MX_IRIGB_NAME);
		retval = -ENOMEM;
		goto cleanup_probe0;
	}

	pIRIG_B->vaddr = ioremap(pIRIG_B->phyaddr, pIRIG_B->phylen);
	reg = pIRIG_B->vaddr;

	if (pci_enable_device(pdev) < 0) {
		printk("pci_enable_device() fail !\n");
		goto cleanup_probe1;
	}

	writel(SYSCON_BIT_RESET, reg+SYSCON); /* Reset this chip */

	/* The IRQ of IRIG-B device */
	pIRIG_B->irq = pdev->irq;
	printk(KERN_INFO"request_irq():%d\n", pIRIG_B->irq);

	retval = request_irq(pIRIG_B->irq, irigb_interrupt, IRQF_SHARED, MX_IRIGB_NAME, (void *)&IRIG_B);
	if (retval < 0) {
		printk("request_irq(), pIRIG_B->irq:%d, fail !\n", pIRIG_B->irq);
		goto cleanup_probe2;
	}


#ifdef DEBUG	/* Just used for debugging */
	unsigned int *reg = pIRIG_B->vaddr;

	/* writel(SYSCON_BIT_RESET, reg+SYSCON ); */ /* Reset this chip */
	printk("reg:%x\n", reg+DEVICEID);
	printk("device id pIRIG_B->vaddr+DEVICEID:%x=%x\n", reg+DEVICEID, readl(reg+DEVICEID));
	printk("date code reg+DATECODE:%x=%x\n", reg+DATECODE, readl(reg+DATECODE));
	printk("date code reg+PORTDAT:%x=%x\n", reg+PORTDAT, readl(reg+PORTDAT));
	printk("INTMASK:%x\n", readl(reg+INTMSK));
	printk("RTCDATA0:%x\n", readl(reg+RTCDAT0));
	printk("RTCDATA1:%x!\n", readl(reg+RTCDAT1));
	printk("RTCDATA2:%x!\n", readl(reg+RTCDAT2));
	printk("RTCDATA3:%x!\n", readl(reg+RTCDAT3));
#endif

	/* Enable Interrupt */
	intMsk &= ~(INTSTS_BIT_IRIG0DE_OFF | \
			INTSTS_BIT_IRIG0DE_FRMERR | \
			INTSTS_BIT_IRIG0DE_PARERR | \
			INTSTS_BIT_IRIG0DE_DONE | \
			INTSTS_BIT_IRIG1DE_OFF | \
			INTSTS_BIT_IRIG1DE_FRMERR | \
			INTSTS_BIT_IRIG1DE_PARERR | \
			INTSTS_BIT_IRIG1DE_DONE | \
			INTSTS_BIT_PPSDE_TIMEOUT | \
			INTSTS_BIT_PPSDE_DONE);

	writel(intMsk, reg + INTMSK);

	pIRIG_B->irigb_misc.minor = MX_IRIGB_MINOR;
	pIRIG_B->irigb_misc.fops = &irigb_fops;
	pIRIG_B->irigb_misc.name = MX_IRIGB_NAME;

	if (misc_register(&pIRIG_B->irigb_misc) < 0) {
		printk("misc_register() fail\n");
		goto cleanup_probe3;
	}

	return 0;

cleanup_probe3:
	/* Mask all the interrupts */
	writel(~0x0, reg+INTMSK);
	/* Free IRQ */
	free_irq(pIRIG_B->irq, &IRIG_B);
cleanup_probe2:
	pci_disable_device(pdev);
cleanup_probe1:
	release_mem_region(pIRIG_B->phyaddr, pIRIG_B->phylen);
	iounmap(pIRIG_B->vaddr);
cleanup_probe0:
	return retval;
}

void irigb_remove(struct pci_dev *pdev)
{
	struct irigb_struct *pIRIG_B;

	printk(KERN_INFO"Remove device %x:%x\n", pdev->vendor, pdev->device);

	pIRIG_B = pci_get_drvdata(pdev);

	/* Mask all the interrupts */
	writel(~0x0, pIRIG_B->vaddr+INTMSK);
	mdelay(10);
	/* Free IRQ */
	free_irq(pIRIG_B->irq, &IRIG_B);

	pci_disable_device(pdev);

	release_mem_region(pIRIG_B->phyaddr, pIRIG_B->phylen);
	iounmap(pIRIG_B->vaddr);

	misc_deregister(&pIRIG_B->irigb_misc);
}

static int __init irigb_init(void)
{
	int retval = 0;

	printk(KERN_INFO"%s Driver version %s\n", MX_IRIGB_NAME, DRIVER_VERSION);

	retval = pci_register_driver(&irigb_driver);
	if (retval < 0) {
		printk(KERN_ERR "Can't register pci driver\n");
		return -ENODEV;
	}

	return 0;
}

void __exit irigb_exit(void)
{
	printk(KERN_INFO"%s removed\n", MX_IRIGB_NAME);
	pci_unregister_driver(&irigb_driver);
}

module_init(irigb_init);
module_exit(irigb_exit);

MODULE_LICENSE("Proprietary");
MODULE_AUTHOR("jared.wu@moxa.com");
MODULE_DESCRIPTION("IRIG-B module device driver");



moxa_irigb.h
#ifndef __IRIGB__
#define __IRIGB__

#define MX_IRIGB_MINOR	111
#define MX_IRIGB_NAME	"moxa_irigb"

#define IOCTL_GET_REGISTER		_IOR(MX_IRIGB_MINOR,0x800,unsigned long)
#define IOCTL_SET_REGISTER		_IOW(MX_IRIGB_MINOR,0x801,unsigned long)
#define IOCTL_SETCLR_REGISTER_BIT	_IOW(MX_IRIGB_MINOR,0x802,unsigned long)
#define IOCTL_GET_TIMESRC_STATUS	_IOW(MX_IRIGB_MINOR,0x803,unsigned long)

#define MAX_PAIRS 4

struct reg_val_pair_struct {
	u64 count;
	u64 addr[MAX_PAIRS];
	u64 val[MAX_PAIRS];
};

struct reg_bit_pair_struct {
	u64 addr;
	u64 set_bit;
	u64 clear_bit;
};

//-----------------------------------------------------------------------------
// PCI Device/Vendor Ids.
//-----------------------------------------------------------------------------

#ifndef PCI_VENDOR_ID_MOXA
#define	PCI_VENDOR_ID_MOXA	0x1393	/* Moxa vendor ID */
#endif

#ifndef MX_IRIGB_DEVICE_ID
#define MX_IRIGB_DEVICE_ID      0xA004
#endif

//-----------------------------------------------------------------------------
// Define the Registers Offset
//-----------------------------------------------------------------------------
enum _FPGA_REGS {
	DEVICEID,		// Device ID Register
	DATECODE,		// FPGA firmware date code
	SYSCON,			// System configuration
	LPBTCNT,		// Loop back test counter
	INPORTCON,		// Input port configuration
	OUTPORTCON,		// Output port configuration
	PORTDAT,		// Input / Output port data
	IRIGBDE0DAT0,		// IRIG-B Decoder 0 Data 0
	IRIGBDE0DAT1,		// IRIG-B Decoder 0 Data 1
	IRIGBDE0DAT2,		// IRIG-B Decoder 0 Data 2
	IRIGBDE0DAT3,		// IRIG-B Decoder 0 Data 3
	IRIGBDE0CNT,		// IRIG-B Decoder 0 Counter, count start from 2 'P' bit
	IRIGBDE1DAT0,		// IRIG-B Decoder 1 Data 0
	IRIGBDE1DAT1,		// IRIG-B Decoder 1 Data 1
	IRIGBDE1DAT2,		// IRIG-B Decoder 1 Data 2
	IRIGBDE1DAT3,		// IRIG-B Decoder 1 Data 3
	IRIGBDE1CNT,		// IRIG-B Decoder 1 Counter, count start from 2 'P' bit
	PPSCON,			// Pulse per second configuration
	TMCON,			// Time module configuration
	INTSTS,			// Interrupt status
	INTMSK,			// Interrupt mask
	PPSDETIMEOUT,		// Pulse per second timeout
	RTCCON,			// RTC Configuration
	RTCDAT0,		// RTC Data register 0
	RTCDAT1,		// RTC Data register 1
	RTCDAT2,		// RTC Data register 2
	RTCDAT3,		// RTC Data register 3
	RTCLS,			// RTC Leap second
	RTCDST,			// RTC DST
	SFIFOCON,		// Serial FIFO configuration
	SFIFODAT,		// Serial FIFO data
	NLEDCON,		// Notify LED configuration

	MAX_ITEMS
};

//-----------------------------------------------------------------------------
// Define the System Configuration Registers
//-----------------------------------------------------------------------------
#define SYSCON_BIT_RESET			(1<<0)
#define SYSCON_BIT_LBT				(1<<31)

//-----------------------------------------------------------------------------
// Define the Loop back test counter Registers
//-----------------------------------------------------------------------------
#define LPBTCNT_UNIT    			40 /* 40 nano second per unit */

//-----------------------------------------------------------------------------
// Define the input port configuration Registers
//-----------------------------------------------------------------------------
#define INPSEL_INP0     			0
#define INPSEL_INP1     			1
#define INPSEL_INP2     			2
#define INPSEL_INP3     			3
#define INPSEL_INP4     			4
#define INPSEL_INP5     			5
#define INPSEL_INP6     			6
#define INPSEL_INP7     			7

#define INPORTCON_MASK				0x7
#define INPORTCON_IRIGDE0_BIT_S			0
#define INPORTCON_IRIGDE1_BIT_S			4
#define INPORTCON_PPSDE_BIT_S			8
#define INPORTCON_LBT_BIT_S			12
#define INPORTCON_BIT_IRIGDE0_DIS		(1<<3)
#define INPORTCON_BIT_IRIGDE1_DIS		(1<<7)
#define INPORTCON_BIT_PPSDE_DIS			(1<<11)
#define INPORTCON_BIT_LBT_DIS			(1<<15)
#define INPORTCON_BIT_INV0			(1<16)
#define INPORTCON_BIT_INV1			(1<17)
#define INPORTCON_BIT_INV2			(1<18)
#define INPORTCON_BIT_INV3			(1<19)
#define INPORTCON_BIT_INV4			(1<20)
#define INPORTCON_BIT_INV5			(1<21)
#define INPORTCON_BIT_INV6			(1<22)
#define INPORTCON_BIT_INV7			(1<23)

//-----------------------------------------------------------------------------
// Define the output port configuration Registers
//-----------------------------------------------------------------------------
#define OUTPSEL_GPO         			0x0
#define OUTPSEL_IRIGBEN     			0x1
#define OUTPSEL_PPSEN       			0x3
#define OUTPSEL_HIZ         			0x5
#define OUTPSEL_SOUT        			0x6
#define OUTPSEL_LPTS        			0x7
#define OUTPSEL_INP0        			0x8
#define OUTPSEL_INP1        			0x9
#define OUTPSEL_INP2        			0xa
#define OUTPSEL_INP3        			0xb
#define OUTPSEL_INP4        			0xc
#define OUTPSEL_INP5        			0xd
#define OUTPSEL_INP6        			0xe
#define OUTPSEL_INP7        			0xf

#define OUTPORTCON_MASK				0xf
#define OUTPORTCON_P0_BIT_S			0
#define OUTPORTCON_P1_BIT_S			5
#define OUTPORTCON_P2_BIT_S			10
#define OUTPORTCON_P3_BIT_S			15
#define OUTPORTCON_P4_BIT_S			20
#define OUTPORTCON_P5_BIT_S			25

#define OUTPORTCON_BIT_INV0			(1<<4)
#define OUTPORTCON_BIT_INV1			(1<<9)
#define OUTPORTCON_BIT_INV2			(1<<14)
#define OUTPORTCON_BIT_INV3			(1<<19)
#define OUTPORTCON_BIT_INV4			(1<<24)
#define OUTPORTCON_BIT_INV5			(1<<29)
#define OUTPORTCON_BIT_INV6			(1<<30)
#define OUTPORTCON_BIT_INV7			(1<<31)

//-----------------------------------------------------------------------------
// Define the port data Registers
//-----------------------------------------------------------------------------
#define PORTDATA_MASK				(0xFFFF)
#define PORTDATA_INPUT_BIT_S			(0)
#define PORTDATA_OUTPUT_BIT_S			(16)

//-----------------------------------------------------------------------------
// Define the IRIG-B decode data Registers
//-----------------------------------------------------------------------------
#define IRIGBDEDAT0_SEC_MASK			(0xf)
#define IRIGBDEDAT0_SEC_BIT_S			(0)
#define IRIGBDEDAT0_TSEC_MASK			(0x7)
#define IRIGBDEDAT0_TSEC_BIT_S			(5)
#define IRIGBDEDAT0_MIN_MASK			(0xf)
#define IRIGBDEDAT0_MIN_BIT_S			(8)
#define IRIGBDEDAT0_TMIN_MASK			(0x7)
#define IRIGBDEDAT0_TMIN_BIT_S			(13)
#define IRIGBDEDAT0_HOUR_MASK			(0xf)
#define IRIGBDEDAT0_HOUR_BIT_S			(17)
#define IRIGBDEDAT0_THOUR_MASK			(0x3)
#define IRIGBDEDAT0_THOUR_BIT_S			(22)

#define IRIGBDEDAT1_DAY_MASK			(0xf)
#define IRIGBDEDAT1_DAY_BIT_S			(0)
#define IRIGBDEDAT1_TDAY_MASK			(0x3f)
#define IRIGBDEDAT1_TDAY_BIT_S			(5)
#define IRIGBDEDAT1_HDAY_MASK			(0x3)
#define IRIGBDEDAT1_HDAY_BIT_S			(9)
#define IRIGBDEDAT1_YEAR_MASK			(0xf)
#define IRIGBDEDAT1_YEAR_BIT_S			(18)
#define IRIGBDEDAT1_TYEAR_MASK			(0xf)
#define IRIGBDEDAT1_TYEAR_BIT_S			(23)

#define IRIGBDEDAT2_BIT_LSP			(1<<0)
#define IRIGBDEDAT2_BIT_LS			(1<<1)
#define IRIGBDEDAT2_BIT_DSP			(1<<2)
#define IRIGBDEDAT2_BIT_DST			(1<<3)
#define IRIGBDEDAT2_BIT_TZS			(1<<4)
#define IRIGBDEDAT2_BIT_TZH			(1<<9)
#define IRIGBDEDAT2_TZ_MASK			(0xf)
#define IRIGBDEDAT2_TZ_BIT_S			(5)
#define IRIGBDEDAT2_TQ_MASK			(0xf)
#define IRIGBDEDAT2_TQ_BIT_S			(10)
#define IRIGBDEDAT2_BIT_PAR			(1<<14)

//-----------------------------------------------------------------------------
// Define the Pulse per second configuration Registers
//-----------------------------------------------------------------------------
#define PPSCON_DE_HOLDTIME_MASK			(0xffff)
#define PPSCON_DE_HOLDTIME_BIT_S		(0)
#define PPSCON_EN_PULSEWIDTH_MASK		(0xffff)
#define PPSCON_EN_PULSEWIDTH_BIT_S		(16)

//-----------------------------------------------------------------------------
// Define the Time module configuration Registers
//-----------------------------------------------------------------------------
#define TMCON_BIT_IRIGDE0PARCHK			(1<<0)
#define TMCON_BIT_IRIGDE1PARCHK			(1<<1)
#define TMCON_BIT_IRIGENPARCHK			(1<<2)

//-----------------------------------------------------------------------------
// Define the Interrupt status/mask Registers
//-----------------------------------------------------------------------------
#define INTSTS_BIT_IRIG0DE_OFF			(1<<0)
#define INTSTS_BIT_IRIG0DE_FRMERR		(1<<1)
#define INTSTS_BIT_IRIG0DE_PARERR		(1<<2)
#define INTSTS_BIT_IRIG0DE_DONE			(1<<3)
#define INTSTS_BIT_IRIG1DE_OFF			(1<<4)
#define INTSTS_BIT_IRIG1DE_FRMERR		(1<<5)
#define INTSTS_BIT_IRIG1DE_PARERR		(1<<6)
#define INTSTS_BIT_IRIG1DE_DONE			(1<<7)
#define INTSTS_BIT_PPSDE_TIMEOUT		(1<<8)
#define INTSTS_BIT_PPSDE_DONE			(1<<9)
#define INTSTS_BIT_IRIGEN_DONE			(1<<10)
#define INTSTS_BIT_PPSEN_DONE			(1<<11)

//-----------------------------------------------------------------------------
// Define the RTC configuration Registers
//-----------------------------------------------------------------------------
#define RTCCON_SYNCSRC_MASK			(0x7)
#define RTCCON_SYNCSRC_FREERUN			(0x0)
#define RTCCON_SYNCSRC_IRIG0			(0x1)
#define RTCCON_SYNCSRC_IRIG1			(0x2)
#define RTCCON_SYNCSRC_PPS			(0x3)
#define RTCCON_BIT_LS_TYPE			(1<<5)

//-----------------------------------------------------------------------------
// Define the RTC data Registers
//-----------------------------------------------------------------------------
#define RTCDAT0_SEC_MASK			(0xff)
#define RTCDAT0_SEC_BIT_S			(0)
#define RTCDAT0_MIN_MASK			(0xff)
#define RTCDAT0_MIN_BIT_S			(8)
#define RTCDAT0_HOUR_MASK			(0xff)
#define RTCDAT0_HOUR_BIT_S			(16)
#define RTCDAT0_DAY_MASK			(0xff)
#define RTCDAT0_DAY_BIT_S			(24)
#define RTCDAT1_MONTH_MASK			(0xff)
#define RTCDAT1_MONTH_BIT_S			(0)
#define RTCDAT1_YEAR_MASK			(0xffff)
#define RTCDAT1_YEAR_BIT_S			(8)
#define RTCDAT1_BIT_COMMIT_TIME			(1<<31)
#define RTCDAT2_NANO_SEC
#define RTCDAT3_BIT_LSP				(1<<0)
#define RTCDAT3_BIT_LS				(1<<1)
#define RTCDAT3_BIT_DSP				(1<<2)
#define RTCDAT3_BIT_DST				(1<<3)
#define RTCDAT3_BIT_TZS				(1<<4)
#define RTCDAT3_BIT_TZH				(1<<5)
#define RTCDAT3_TZ_MASK				(0xf)
#define RTCDAT3_TZ_BIT_S			(6)
#define RTCDAT3_TQ_MASK				(0xf)
#define RTCDAT3_TQ_BIT_S			(10)

//-----------------------------------------------------------------------------
// Define the RTC leap second Registers
//-----------------------------------------------------------------------------
#define RTCLS_MIN_MASK				(0x7f)
#define RTCLS_MIN_BIT_S				(0)
#define RTCLS_HOUR_MASK				(0x3f)
#define RTCLS_HOUR_BIT_S			(7)
#define RTCLS_DAY_MASK				(0x3ff)
#define RTCLS_DAY_BIT_S				(13)
#define RTCLS_YEAR_MASK				(0xff)
#define RTCLS_YEAR_BIT_S			(23)
#define RTCLS_BIT_ENABLE			(1<<31)

//-----------------------------------------------------------------------------
// Define the RTC daylight saving time Registers
//-----------------------------------------------------------------------------
#define RTCDST_ST_HOUR_MASK			(0x3f)
#define RTCDST_ST_HOUR_BIT_S			(0)
#define RTCDST_ST_DAY_MASK			(0x3ff)
#define RTCDST_ST_DAY_BIT_S			(6)
#define RTCDST_ED_HOUR_MASK			(0x3f)
#define RTCDST_ED_HOUR_BIT_S			(16)
#define RTCDST_ED_DAY_MASK			(0x3ff)
#define RTCDST_ED_DAY_BIT_S			(22)

//-----------------------------------------------------------------------------
// Define the Serial FIFO configuration Registers
//-----------------------------------------------------------------------------
#define SFIFO_BIT_STS_FULL			(1<<16)
#define SFIFO_BIT_STS_EMPTY			(1<<17)
#define SFIFO_BIT_STS_CLEAR			(1<<18)

//-----------------------------------------------------------------------------
// Define the Serial FIFO data Registers
//-----------------------------------------------------------------------------
#define SFIFODAT_WIDTH_MASK			(0x7fffffff)
#define SFIFODAT_WIDTH_BIT_S			(0)
#define SFIFO_BIT_LEVEL				(1<<31)

//-----------------------------------------------------------------------------
// Define the Notify LED configuration Registers
//-----------------------------------------------------------------------------
#define NLED_MODE_OUTP0				(0)
#define NLED_MODE_OUTP1				(1)
#define NLED_MODE_OUTP2				(2)
#define NLED_MODE_OUTP3				(3)
#define NLED_MODE_OUTP4				(4)
#define NLED_MODE_OUTP5				(5)
#define NLED_MODE_IRIG0_OK			(6)
#define NLED_MODE_IRIG1_OK			(7)
#define NLED_MODE_INP0				(8)
#define NLED_MODE_INP1				(9)
#define NLED_MODE_INP2				(10)
#define NLED_MODE_INP3				(11)
#define NLED_MODE_INP4				(12)
#define NLED_MODE_INP5				(13)
#define NLED_MODE_INP6				(14)
#define NLED_MODE_INP7				(15)
#define NLED_MODE_MASK				(0xf)
#define NLED_P0_BIT_S				(0)
#define NLED_P1_BIT_S				(4)
#define NLED_P2_BIT_S				(8)
#define NLED_P3_BIT_S				(12)
#define NLED_P4_BIT_S				(16)
#define NLED_P5_BIT_S				(20)
#define NLED_P6_BIT_S				(24)
#define NLED_P7_BIT_S				(28)


#endif /* endof __IRIGB__ */
